Sources/Plasma/FeatureLib/pfMetalPipeline/plMetalDevice.cpp

```diff
@@ -0,0 +1,1279 @@
+}
+
+void plMetalDevice::StartPipelineBuild(plMetalPipelineRecord& record, std::condition_variable** condOut)
+{
+    __block std::condition_variable* newCondition = new std::condition_variable();
+    fConditionMap[record] = newCondition;
+    if (condOut) {
+        *condOut = newCondition;
+    }
+
+    if (fNewPipelineStateMap[record] != NULL) {
+        return fNewPipelineStateMap[record];
+    }
+
+    MTL::Library* library = fMetalDevice->newDefaultLibrary();
+
+    std::shared_ptr<plMetalPipelineState> pipelineState = record.state;
+
+    MTL::RenderPipelineDescriptor* descriptor = MTL::RenderPipelineDescriptor::alloc()->init();
+    descriptor->setLabel(pipelineState->GetDescription());
+
+    const MTL::Function* vertexFunction = pipelineState->GetVertexFunction(library);
+    const MTL::Function* fragmentFunction = pipelineState->GetFragmentFunction(library);
+    descriptor->setVertexFunction(vertexFunction);
+    descriptor->setFragmentFunction(fragmentFunction);
+
+    descriptor->colorAttachments()->object(0)->setBlendingEnabled(true);
+    pipelineState->ConfigureBlend(descriptor->colorAttachments()->object(0));
+
+    MTL::VertexDescriptor* vertexDescriptor = MTL::VertexDescriptor::vertexDescriptor();
+    pipelineState->ConfigureVertexDescriptor(vertexDescriptor);
+    descriptor->setVertexDescriptor(vertexDescriptor);
+    descriptor->setDepthAttachmentPixelFormat(record.depthFormat);
+    descriptor->colorAttachments()->object(0)->setPixelFormat(record.colorFormat);
+
+    descriptor->setSampleCount(record.sampleCount);
+
+    NS::Error* error;
+    fMetalDevice->newRenderPipelineState(descriptor, ^(MTL::RenderPipelineState* pipelineState, NS::Error* error) {
+        if (error) {
+            // leave the condition in place for now, we don't want to
+            // retry if the shader is defective. the condition will
+            // prevent retries
+            hsAssert(0, error->localizedDescription()->cString(NS::UTF8StringEncoding));
+        } else {
+            plMetalLinkedPipeline* linkedPipeline = new plMetalLinkedPipeline();
+            linkedPipeline->pipelineState = pipelineState->retain();
+            linkedPipeline->fragFunction = fragmentFunction;
+            linkedPipeline->vertexFunction = vertexFunction;
+
+            fNewPipelineStateMap[record] = linkedPipeline;
+            // signal that we're done
+            newCondition->notify_all();
+        }
+    });
+
+    descriptor->release();
+    library->release();
+}
+
+plMetalDevice::plMetalLinkedPipeline* plMetalDevice::PipelineState(plMetalPipelineState* pipelineState)
+{
+    MTL::PixelFormat depthFormat = fCurrentDepthFormat;
+    MTL::PixelFormat colorFormat = fCurrentFragmentOutputTexture->pixelFormat();
+
+    plMetalPipelineRecord record = {
+        depthFormat,
+        colorFormat,
+        CurrentTargetSampleCount()};
+
+    record.state = std::shared_ptr<plMetalPipelineState>(pipelineState->Clone());
```

Comment:
I don't know. I'll preface by saying I'm not happy with how plMetalPipelineState turned out.

This is a heavily hit subsystem - and so the goal is to prevent copies of plMetalPipelineState objects. But you have to create a plMetalPipelineState to query the system.

plMetalPipelineState works - and is stable. But it's extremely brittle and needs a deeper reconsider - which the use of shader_ptr falls under.